### PR TITLE
Hotfix: move CUDAHOSTCXX

### LINKED ISSUE #17823: Cray environment for CUDAHOSTCXX is goofed
### Steps to reproduce the issue

On CoriGPU I am building lbann using GCC 8.2.0 and cmake is unable to resolve CUDAHOSTCXX because spack setups up a cray environment even though I am not using one.  It ends up with CUDAHOSTCXX=CC, which is wrong.

### Error Message

```console
CMake Error at /global/u2/v/vanessen/spack.git/opt/spack/cray-cnl7-skylake_avx512/gcc-8.2.0/cmake-3.18.0-esuryqzyhasr24bju42ydqhlnvlvmd4b/share/cmake-3.18/Modules/CMakeDetermineCUDACompiler.cmake:46 (message):
  Could not find compiler set in environment variable CUDAHOSTCXX:
  CC.
Call Stack (most recent call first):
  CMakeLists.txt:264 (enable_language)
```

```console
[lbann-dev-skylake_avx512] vanessen@cgpu12:~/lbann.git/build/gnu.Release.cgpu.nersc.gov/build> printenv CUDAHOSTCXX
CC
```

### Information on your system

CoriGPU

### Additional information

- [ ] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [ ] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by ax3l**
Fixed #17823.

Set only in dependent packages.

@bvanessen I am just starting my PTO, can you test if this works for you, please?

Maintainers: please feel free to push, close, reopen this PR as needed. cc @svenevs @robo-wylder

Regression of #16869

### CODE REVIEW

File: var/spack/repos/builtin/packages/cuda/package.py [hunk 1]
```diff
@@ -80,15 +80,16 @@ class Cuda(Package):
...
 
     def setup_build_environment(self, env):
-        env.set('CUDAHOSTCXX', self.compiler.cxx)
         if self.spec.satisfies('@10.1.243:'):
             libxml2_home  = self.spec['libxml2'].prefix
             env.set('LIBXML2HOME', libxml2_home)
             env.append_path('LD_LIBRARY_PATH', libxml2_home.lib)
 
+    def setup_dependent_build_environment(self, env, dependent_spec):
+        env.set('CUDAHOSTCXX', self.compiler.cxx)
```
Comment (ax3l) [inline var/spack/repos/builtin/packages/cuda/package.py]: should probably be the compiler of the dependent spec?

Comment (tgamblin) [reply var/spack/repos/builtin/packages/cuda/package.py]: yes:

```python
dependent_spec.package.compiler.cxx
```

**Comment by tgamblin** [comment]
@bvanessen: please test

**Comment by bvanessen** [comment]
@tgamblin I am checking right now

**Comment by bvanessen** [comment]
This PR fixes the problem for me.

File: var/spack/repos/builtin/packages/cuda/package.py (hunk 1 above)
Comment (ax3l) [reply var/spack/repos/builtin/packages/cuda/package.py]:
```suggestion
        env.set('CUDAHOSTCXX', dependent_spec.package.compiler.cxx)
```

**Review by svenevs — APPROVED** [review]
Seems like the right choice to me.  I don't see any reason this would cause unforseen problems, the concretizer will make it difficult for people to accidentally mix e.g., clang hostcxx vs gcc. AFAICT you'd have to deliberately force spack via "invalid configs" to end up mixing different hostcxx in a way that's problematic (because we're shadowing CXX).

If we encounter additional problems in the future we can also consider just dropping it. I prefer setting it explicitly as we are here, but in practice the rest of the spack build env  (setting CXX) means CMake will fallback to the right thing rather than searching for something unexpected.

Thanks for fixing / testing!

**Comment by tgamblin** [comment]
@bvanessen: can you try one more time with @ax3l’s latest commit?  If you can build the whole LBANN stack with it (not just LBANN in an env) that would be good to know.

**Review by adamjstewart — APPROVED** [review]

**Comment by bvanessen** [comment]
@tgamblin @ax3l This works for me both with a source build and a full spack build.

**Pull request merged**